### PR TITLE
Add FuEfiFirmware to parse EFI file volumes

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [X] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by hughsie** [comment]
Other things pending, e.g. FFS2 -- but I wanted to get the basics in place first.

File: libfwupdplugin/fu-efi-firmware-volume.c [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+	g_autoptr(GBytes) blob = NULL;
+
+	/* check if this is _FIT */
+	if (!fu_common_read_uint32_safe (buf, bufsz, offset,
+					 &sig, G_LITTLE_ENDIAN, error)) {
+		g_prefix_error (error, "failed to read start signature: ");
+		return FALSE;
+	}
+	if (sig == FU_EFI_FIT_SIGNATURE) {
+		*offset_val += 0x150000;
```
Comment (superm1) [inline libfwupdplugin/fu-efi-firmware-volume.c:76]: I think this magic number should be a define.

**Review by superm1 — COMMENTED** [review]
why commit the binary files to tree?  it seems that they can be generated by your XML.

File: libfwupdplugin/fu-efi-firmware-volume.c [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+	for (guint i = 0; i < 0x10; i++) {
+		if (buf[offset + FU_EFI_FIRMWARE_VOLUME_OFFSET_ZERO_VECTOR + i] != 0x0) {
+			g_set_error (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "reserved section invalid @0x%x",
+				     (guint) offset + FU_EFI_FIRMWARE_VOLUME_OFFSET_ZERO_VECTOR + i);
+			return FALSE;
+		}
+	}
```
Comment (superm1) [inline libfwupdplugin/fu-efi-firmware-volume.c:104]: does the spec actually say that it should be zero?  Usually reserved means don't care - it should be zero, but a future version of a spec can encroach into this area and so looking for zero might mean a failure later.

File: libfwupdplugin/fu-efi-firmware-volume.c [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "invalid volume length");
+		return FALSE;
+	}
+	if (!fu_common_read_uint32_safe (buf, bufsz,
+					 offset + FU_EFI_FIRMWARE_VOLUME_OFFSET_ATTRS,
+					 &attrs, G_LITTLE_ENDIAN, error))
+		return FALSE;
```
Comment (superm1) [inline libfwupdplugin/fu-efi-firmware-volume.c:133]: don't you want to prefix the error here?

Comment (hughsie) [reply libfwupdplugin/fu-efi-firmware-volume.c:133]: Agree, fixed.

File: libfwupdplugin/fu-efi-firmware-volume.c (hunk 2 above)
Comment (hughsie) [reply libfwupdplugin/fu-efi-firmware-volume.c:104]: You're completely correct; Apple firmware seems to put some kind of CRC here. I'll remove this check.

File: libfwupdplugin/fu-efi-firmware-volume.c (hunk 1 above)
Comment (hughsie) [reply libfwupdplugin/fu-efi-firmware-volume.c:76]: Agree, fixed.

**Comment by hughsie** [comment]
> why commit the binary files to tree

Good question; the oss-fuzz.py doesn't actually do a build, and so the generated files wouldn't be created. I do wonder if that should be created using a build-fuzzing-samples.sh or something, and that can be called from meson and from the oss-fuzz.py thing.

**Comment by hughsie** [comment]
I'll pull the FFS2 stuff out, that's not baked very well yet.

**Comment by superm1** [comment]
> Good question; the oss-fuzz.py doesn't actually do a build, and so the generated files wouldn't be created. I do wonder if that should be created using a build-fuzzing-samples.sh or something, and that can be called from meson and from the oss-fuzz.py thing.

Yes, I think having a separate build target of some sort to make them makes most sense.

File: libfwupdplugin/fu-efi-firmware-volume.c [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+	if (!fu_common_read_uint16_safe (buf, bufsz,
+					 offset + FU_EFI_FIRMWARE_VOLUME_OFFSET_EXT_HDR,
+					 &ext_hdr, G_LITTLE_ENDIAN, error)) {
+		g_prefix_error (error, "failed to read ext_hdr: ");
+		return FALSE;
+	}
+	if (!fu_common_read_uint8_safe (buf, bufsz,
+					offset + FU_EFI_FIRMWARE_VOLUME_OFFSET_REVISION,
+					&revision, error))
+		return FALSE;
```
Comment (superm1) [inline libfwupdplugin/fu-efi-firmware-volume.c:153]: missing a prefix error here too

**Comment by hughsie** [comment]
Galaxy brain wave: I don't like the design this headed in: The FuEfiFirmwareVolume should be parsing "one thing" and not adding the images as different volumes. It's kinda working around that there doesn't exist a FuBiosFirmware that reads in the BIOS section of the IFD and spits out the different FVs. At the moment I only care about the FVs in the BIOS image, but what if we did actually want to look at the _FIT_ in the future. New patch incoming, probably tomorrow.

**Comment by superm1** [comment]
> FuBiosFirmware that reads in the BIOS section of the IFD and spits out the different FVs

Well not to mention that you can actually have UEFI and firmware volumes on a lot of architectures that aren't Intel.

**Comment by hughsie** [comment]
> you can actually have UEFI and firmware volumes on a lot of architectures that aren't Intel

But isn't the BIOS blob in the IFD Intel specific? e.g. _jump 16MiB, read FV, read FV, read FV, read _FIT__ seems really Intel specific to me.

**Comment by superm1** [comment]
> > you can actually have UEFI and firmware volumes on a lot of architectures that aren't Intel
> 
> But isn't the BIOS blob in the IFD Intel specific? e.g. _jump 16MiB, read FV, read FV, read FV, read _FIT__ seems really Intel specific to me.

Yes, but this PR is about parsing EFI FV's.  I'm in agreement I think you need a level of abstraction up, but you need one in the middle too.  Maybe something like this:

FuPlatformFw
-> FuIntelFIT

Then you have an opportunity to have alternatives to `FuIntelFit`.  And multiple types can reference `FuEfiVolume`.

**Pull request closed without merging**